File: server/lib/comparison_service.rb
require_relative 'worker_pool'
require_relative 'indexing_service'
require_relative 'dimension_filter'
require_relative 'image_comparator'

class ComparisonService
  BATCH_SIZE = 100  # 每批次插入100条记录

  def initialize(project)
    @project = project
    @worker_pool = WorkerPool.new(4)
    @candidate_batch = []
    @update_batch = []
    @mutex = Mutex.new
  end

  def process
    @project.update(status: 'processing', started_at: Time.now)

    begin
      # 第一轮：索引（如果未完成）
      unless ['indexed', 'comparing', 'completed'].include?(@project.status)
        puts "Starting indexing phase..."
        IndexingService.new(@project).process
        puts "Indexing phase completed"
      end

      # 第二轮：匹配
      @project.update(status: 'comparing')
      puts "Starting comparison phase..."

      @worker_pool.start

      # 并行处理source文件
      source_files = @project.source_files.where(status: 'indexed').to_a
      puts "Comparing #{source_files.size} source files..."

      source_files.each do |sf|
        @worker_pool.add_job do
          compare_single_source(sf)
        end
      end

      @worker_pool.stop

      # 批量插入剩余的候选项
      flush_candidate_batch

      # 批量更新剩余的 source_files 状态
      flush_update_batch

      # 自动创建默认选择（选择所有 rank=1 的候选项）
      create_auto_selections

      @project.update(status: 'completed', ended_at: Time.now)
      puts "Comparison phase completed"
    rescue => e
      @project.update(status: 'error', error_message: e.message)
      puts "ComparisonService error: #{e.message}"
      puts e.backtrace
    ensure
      @worker_pool&.stop
    end
  end

  private

  def compare_single_source(source_file)
    @project.project_targets.each do |target|
      # 获取所有target文件（避免 N+1 查询，在外层已经 includes）
      all_targets = target.target_files.to_a

      # 第一步：自适应尺寸筛选
      filtered_targets = DimensionFilter.adaptive_filter_targets(source_file, all_targets)

      puts "Source #{source_file.relative_path}: #{all_targets.size} targets -> #{filtered_targets.size} after adaptive dimension filter"

      # 第二步：哈希比较，使用自适应阈值
      candidates = []

      filtered_targets.each do |tf|
        begin
          # 使用预计算的哈希值
          similarity = calculate_similarity_from_hashes(source_file, tf)

          # 先收集所有相似度数据
          candidates << {
            target_file: tf,
            similarity: similarity
          }
        rescue => e
          puts "Error comparing with target #{tf.id}: #{e.message}"
          next
        end
      end

      # 排序
      candidates.sort_by! { |c| -c[:similarity] }

      # 自适应阈值策略：确保至少有一个候选项
      similarity_thresholds = [50.0, 40.0, 30.0, 20.0, 10.0, 0.0]
      final_candidates = []
      used_threshold = 50.0

      similarity_thresholds.each do |threshold|
        filtered = candidates.select { |c| c[:similarity] > threshold }

        if filtered.any?
          # 限制最多20个候选项
          final_candidates = filtered.first(20)
          used_threshold = threshold
          break
        end
      end

      # 如果所有阈值都没有结果，强制选择相似度最高的1个
      if final_candidates.empty? && candidates.any?
        final_candidates = [candidates.first]
        used_threshold = 0.0
        puts "  强制选择最高相似度候选项: #{candidates.first[:similarity].round(2)}%"
      end

      # 准备批量插入数据
      final_candidates.each_with_index do |cand, index|
        record = {
          source_file_id: source_file.id,
          project_target_id: target.id,
          file_path: cand[:target_file].full_path,
          similarity_score: cand[:similarity],
          rank: index + 1,
          width: cand[:target_file].width,
          height: cand[:target_file].height
        }

        # 线程安全地添加到批次
        @mutex.synchronize do
          @candidate_batch << record

          # 当批次达到阈值时，批量插入
          if @candidate_batch.size >= BATCH_SIZE
            flush_candidate_batch
          end
        end
      end

      puts "Found #{final_candidates.size} candidates for #{source_file.relative_path} in #{target.name} (threshold: #{used_threshold}%)"

      # 及时释放 candidates 数组的内存引用
      candidates = nil
      all_targets = nil
      filtered_targets = nil
    end

    # 标记为已分析（收集到批次中稍后批量更新）
    @mutex.synchronize do
      @update_batch << source_file.id

      if @update_batch.size >= BATCH_SIZE
        flush_update_batch
      end
    end
  rescue => e
    puts "Error comparing source #{source_file.id}: #{e.message}"
    puts e.backtrace
  end

  # 批量插入候选项
  def flush_candidate_batch
    return if @candidate_batch.empty?

    batch = @candidate_batch.dup
    @candidate_batch.clear

    begin
      ComparisonCandidate.insert_all(batch)
      puts "Batch inserted #{batch.size} comparison candidates"
    rescue => e
      puts "Error batch inserting candidates: #{e.message}"
    end
  end

  # 批量更新 source_files 状态
  def flush_update_batch
    return if @update_batch.empty?

    batch = @update_batch.dup
    @update_batch.clear

    begin
      SourceFile.where(id: batch).update_all(status: 'analyzed')
      puts "Batch updated #{batch.size} source files to analyzed"
    rescue => e
      puts "Error batch updating source files: #{e.message}"
    end
  end

  # 自动创建默认选择：为每个 source_file + target 组合选择 rank=1 的候选项
  def create_auto_selections
    puts "Creating auto-selections for best matches..."

    # 获取所有 rank=1 的候选项
    best_candidates = ComparisonCandidate
      .where(rank: 1)
      .joins("INNER JOIN source_files ON source_files.id = comparison_candidates.source_file_id")
      .where("source_files.project_id = ?", @project.id)
      .select(:id, :source_file_id, :project_target_id)

    # 批量创建 TargetSelection 记录
    selections = []
    best_candidates.each do |candidate|
      selections << {
        source_file_id: candidate.source_file_id,
        project_target_id: candidate.project_target_id,
        selected_candidate_id: candidate.id,
        no_match: false,
        created_at: Time.now,
        updated_at: Time.now
      }

      # 批量插入
      if selections.size >= BATCH_SIZE
        TargetSelection.insert_all(selections)
        puts "Batch inserted #{selections.size} auto-selections"
        selections.clear
      end
    end

    # 插入剩余的选择记录
    if selections.any?
      TargetSelection.insert_all(selections)
      puts "Inserted final #{selections.size} auto-selections"
    end

    puts "Auto-selection completed"
  end

  # 使用预计算的哈希值计算相似度
  def calculate_similarity_from_hashes(source, target)
    # Convert string hashes back to integers for comparison
    phash_sim = ImageComparator.hash_similarity(source.phash.to_i, target.phash.to_i, 64)
    ahash_sim = ImageComparator.hash_similarity(source.ahash.to_i, target.ahash.to_i, 1024)
    dhash_sim = ImageComparator.hash_similarity(source.dhash.to_i, target.dhash.to_i, 64)

    source_hist = JSON.parse(source.histogram)
    target_hist = JSON.parse(target.histogram)
    histogram_sim = ImageComparator.histogram_similarity(source_hist, target_hist)

    # 加权平均
    weights = ImageComparator::WEIGHTS
    phash_sim * weights[:phash] +
    ahash_sim * weights[:ahash] +
    dhash_sim * weights[:dhash] +
    histogram_sim * weights[:histogram]
  end
end

File: go-server/internal/services/comparison_service.go
package services

import (
	"context"
	"fmt"
	"log"
	"strconv"
	"sync"
	"time"

	"github.com/bilibili/look-alike/internal/database"
	"github.com/bilibili/look-alike/internal/image"
	"github.com/bilibili/look-alike/internal/models"
)

const (
	batchSize         = 100
	concurrentWorkers = 4
)

// ComparisonService handles comparison of source and target files
type ComparisonService struct {
	project *models.Project
	ctx     context.Context
}

// NewComparisonService creates a new comparison service
func NewComparisonService(project *models.Project, ctx context.Context) *ComparisonService {
	if ctx == nil {
		ctx = context.Background()
	}
	return &ComparisonService{
		project: project,
		ctx:     ctx,
	}
}

// ProcessComparison runs the comparison process
func ProcessComparison(project *models.Project, ctx context.Context) error {
	log.Println("=========================================")
	log.Printf("ComparisonService.Process started for project %d: %s", project.ID, project.Name)
	log.Printf("Project status: %s", project.Status)
	log.Println("=========================================")

	// Update status to processing
	database.DB.Model(&project).Updates(map[string]interface{}{
		"status":     "processing",
		"started_at": time.Now(),
	})

	// Run indexing first if needed
	if project.Status != "indexed" {
		log.Println("[INDEXING] Starting indexing phase...")
		if err := Process(project); err != nil {
			return err
		}
		log.Println("[INDEXING] Indexing phase completed")
	}

	// Validate indexing
	var totalSourceFiles, indexedSourceFiles, totalTargetFiles int64
	database.DB.Model(&models.SourceFile{}).Where("project_id = ?", project.ID).Count(&totalSourceFiles)
	database.DB.Model(&models.SourceFile{}).Where("project_id = ? AND status = ?", project.ID, "indexed").Count(&indexedSourceFiles)

	database.DB.Table("target_files").
		Joins("INNER JOIN project_targets ON project_targets.id = target_files.project_target_id").
		Where("project_targets.project_id = ?", project.ID).
		Count(&totalTargetFiles)

	log.Println("[VALIDATION] Validating indexing results...")
	log.Printf("[VALIDATION] Total source files: %d", totalSourceFiles)
	log.Printf("[VALIDATION] Indexed source files: %d", indexedSourceFiles)
	log.Printf("[VALIDATION] Total target files: %d", totalTargetFiles)

	if totalSourceFiles == 0 {
		err := fmt.Errorf("no source files found after indexing")
		database.DB.Model(&project).Updates(map[string]interface{}{
			"status":        "error",
			"error_message": err.Error(),
		})
		return err
	}

	if totalTargetFiles == 0 {
		err := fmt.Errorf("no target files found after indexing")
		database.DB.Model(&project).Updates(map[string]interface{}{
			"status":        "error",
			"error_message": err.Error(),
		})
		return err
	}

	// Start comparison phase
	log.Println("[COMPARING] Starting comparison phase...")
	database.DB.Model(&project).Update("status", "comparing")

	svc := NewComparisonService(project, ctx)
	if err := svc.compareAll(); err != nil {
		database.DB.Model(&project).Updates(map[string]interface{}{
			"status":        "error",
			"error_message": err.Error(),
		})
		return err
	}

	// Create auto-selections
	if err := svc.createAutoSelections(); err != nil {
		log.Printf("[WARNING] Failed to create auto-selections: %v", err)
	}

	// Update to completed
	database.DB.Model(&project).Updates(map[string]interface{}{
		"status":   "completed",
		"ended_at": time.Now(),
	})

	log.Println("[SUCCESS] Comparison phase completed")
	log.Println("=========================================")
	return nil
}

// compareAll compares all source files with all targets
func (svc *ComparisonService) compareAll() error {
	// Preload all project targets and their files
	var targets []models.ProjectTarget
	if err := database.DB.Preload("TargetFiles").Where("project_id = ?", svc.project.ID).Find(&targets).Error; err != nil {
		return err
	}

	log.Printf("[COMPARING] Preloaded %d targets", len(targets))

	// Get all indexed source files
	var sourceFiles []models.SourceFile
	if err := database.DB.Where("project_id = ? AND status = ?", svc.project.ID, "indexed").Find(&sourceFiles).Error; err != nil {
		return err
	}

	log.Printf("[COMPARING] Processing %d indexed source files...", len(sourceFiles))

	if len(sourceFiles) == 0 {
		return fmt.Errorf("no indexed source files found")
	}

	// Process concurrently with semaphore
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, concurrentWorkers)
	var mu sync.Mutex
	var candidateBatch []models.ComparisonCandidate
	var updateBatch []uint

	for i := range sourceFiles {
		// Check context cancellation
		select {
		case <-svc.ctx.Done():
			log.Println("[COMPARING] Context cancelled, stopping comparison")
			return svc.ctx.Err()
		default:
		}

		wg.Add(1)
		go func(sf *models.SourceFile) {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()

			candidates := svc.compareSingleSource(sf, targets)

			mu.Lock()
			candidateBatch = append(candidateBatch, candidates...)
			updateBatch = append(updateBatch, sf.ID)

			// Batch insert candidates
			if len(candidateBatch) >= batchSize {
				if err := database.DB.Create(&candidateBatch).Error; err != nil {
					log.Printf("[ERROR] Failed to batch insert candidates: %v", err)
				} else {
					log.Printf("[COMPARING] Batch inserted %d candidates", len(candidateBatch))
				}
				candidateBatch = nil
			}

			// Batch update source files
			if len(updateBatch) >= batchSize {
				database.DB.Model(&models.SourceFile{}).Where("id IN ?", updateBatch).Update("status", "analyzed")
				log.Printf("[COMPARING] Batch updated %d source files to analyzed", len(updateBatch))
				updateBatch = nil
			}
			mu.Unlock()
		}(&sourceFiles[i])
	}

	wg.Wait()

	// Insert remaining batches
	if len(candidateBatch) > 0 {
		if err := database.DB.Create(&candidateBatch).Error; err != nil {
			return fmt.Errorf("failed to insert remaining candidates: %w", err)
		}
		log.Printf("[COMPARING] Inserted final batch of %d candidates", len(candidateBatch))
	}

	if len(updateBatch) > 0 {
		database.DB.Model(&models.SourceFile{}).Where("id IN ?", updateBatch).Update("status", "analyzed")
		log.Printf("[COMPARING] Updated final batch of %d source files", len(updateBatch))
	}

	return nil
}

// compareSingleSource compares a single source file with all targets
func (svc *ComparisonService) compareSingleSource(sourceFile *models.SourceFile, targets []models.ProjectTarget) []models.ComparisonCandidate {
	var allCandidates []models.ComparisonCandidate

	for _, target := range targets {
		// Compare with all target files (no dimension filtering)
		log.Printf("Source %s: comparing with %d targets for target %s",
			sourceFile.RelativePath, len(target.TargetFiles), target.Name)

		// Calculate similarities with all targets
		candidates := svc.calculateSimilarities(sourceFile, target.TargetFiles, target.ID)

		// Apply adaptive threshold
		finalCandidates := svc.applyAdaptiveThreshold(candidates, sourceFile.ID, sourceFile.RelativePath, target.Name)

		allCandidates = append(allCandidates, finalCandidates...)
	}

	return allCandidates
}

// calculateSimilarities calculates similarity scores
func (svc *ComparisonService) calculateSimilarities(sourceFile *models.SourceFile, targetFiles []models.TargetFile, targetID uint) []candidateScore {
	var candidates []candidateScore

	for _, tf := range targetFiles {
		similarity := calculateSimilarityFromHashes(sourceFile, &tf)
		candidates = append(candidates, candidateScore{
			targetFile: &tf,
			similarity: similarity,
		})
	}

	return candidates
}

type candidateScore struct {
	targetFile *models.TargetFile
	similarity float64
}

// applyAdaptiveThreshold applies adaptive thresholding
func (svc *ComparisonService) applyAdaptiveThreshold(candidates []candidateScore, sourceFileID uint, sourcePath, targetName string) []models.ComparisonCandidate {
	if len(candidates) == 0 {
		return nil
	}

	// Sort by similarity descending
	for i := 0; i < len(candidates)-1; i++ {
		for j := i + 1; j < len(candidates); j++ {
			if candidates[j].similarity > candidates[i].similarity {
				candidates[i], candidates[j] = candidates[j], candidates[i]
			}
		}
	}

	// Try different thresholds
	thresholds := []float64{50.0, 40.0, 30.0, 20.0, 10.0, 0.0}
	var finalCandidates []candidateScore

	for _, threshold := range thresholds {
		filtered := []candidateScore{}
		for _, c := range candidates {
			if c.similarity > threshold {
				filtered = append(filtered, c)
			}
		}

		if len(filtered) > 0 {
			// Limit to 50 candidates
			if len(filtered) > 50 {
				filtered = filtered[:50]
			}
			finalCandidates = filtered
			break
		}
	}

	// Force select top candidate if none found
	if len(finalCandidates) == 0 && len(candidates) > 0 {
		finalCandidates = []candidateScore{candidates[0]}
		log.Printf("  Forced selection of highest similarity candidate: %.2f%%", candidates[0].similarity)
	}

	// Convert to ComparisonCandidate models
	var result []models.ComparisonCandidate
	for i, c := range finalCandidates {
		result = append(result, models.ComparisonCandidate{
			SourceFileID:    sourceFileID,
			ProjectTargetID: c.targetFile.ProjectTargetID,
			FilePath:        c.targetFile.FullPath,
			SimilarityScore: c.similarity,
			Rank:            i + 1,
			Width:           c.targetFile.Width,
			Height:          c.targetFile.Height,
		})
	}

	log.Printf("Found %d candidates for %s in %s", len(result), sourcePath, targetName)
	return result
}

// calculateSimilarityFromHashes calculates similarity using pre-computed phash only
func calculateSimilarityFromHashes(source *models.SourceFile, target *models.TargetFile) float64 {
	// Parse phash from strings
	sourcePhash, _ := strconv.ParseUint(source.Phash, 10, 64)
	targetPhash, _ := strconv.ParseUint(target.Phash, 10, 64)

	// Calculate phash similarity
	phashSim := image.HashSimilarity(sourcePhash, targetPhash, 64)

	return phashSim
}

// createAutoSelections creates default selections for rank 1 candidates
func (svc *ComparisonService) createAutoSelections() error {
	log.Println("Creating auto-selections for best matches...")

	var bestCandidates []models.ComparisonCandidate
	database.DB.Where("rank = ?", 1).
		Joins("INNER JOIN source_files ON source_files.id = comparison_candidates.source_file_id").
		Where("source_files.project_id = ?", svc.project.ID).
		Find(&bestCandidates)

	var selections []models.TargetSelection
	for _, candidate := range bestCandidates {
		selections = append(selections, models.TargetSelection{
			SourceFileID:        candidate.SourceFileID,
			ProjectTargetID:     candidate.ProjectTargetID,
			SelectedCandidateID: &candidate.ID,
			NoMatch:             false,
		})

		if len(selections) >= batchSize {
			if err := database.DB.Create(&selections).Error; err != nil {
				log.Printf("[ERROR] Failed to batch insert selections: %v", err)
			} else {
				log.Printf("Batch inserted %d auto-selections", len(selections))
			}
			selections = nil
		}
	}

	if len(selections) > 0 {
		if err := database.DB.Create(&selections).Error; err != nil {
			return err
		}
		log.Printf("Inserted final %d auto-selections", len(selections))
	}

	log.Println("Auto-selection completed")
	return nil
}
